docs: misc content & guide updates (#2359)
iterative/dvc.org: content/docs/start/data-and-model-access.md

title: 'Get Started: Data and Model Access'
---

> # Data and Model Access

Reviewer: These are by design for SEO.

Author: Are you sure?

- "Get Started" already appears in the `head.title` tag
- it also appears in the containing `nav` section
- does it really need to be in the `body.h1` element too?
- other subsections in the `nav` don't prefix their supersections in their `body.h1`s

is "Get Started" is some sort of SEO exception to the rule?

Reviewer: > "Get Started" already appears in the head.title tag

Yep, I think that was the main reason: to match the HTML title (I guess helps ranking). Page previews on Google or anywhere will not always have the "Get Started" prefix, which can make it hard to clearly differentiate GS content vs. other similarly titled docs like the Versioning use case, Experiments guide, or Pipelines file ref. We want users to click on and land on the most appropriate content (decreases bounce rate).

That's what I remember but it was a while ago we made this decision. In any case if there's no strong reason to change this lets trust our previous choices, I think.

Reviewer: > other subsections in the nav don't prefix their supersections in their body.h1s

The one tutorial we have does. And all the How to pages.

Author: Hmm may have to re-look at this later. If it's a thing, we could automate it rather than duplicate the title in all source files.

Reviewer: Yep, let's roll back and make an issue to look into it?

Author: yes; reverted (forgot to push earlier).

> We've learned how to _track_ data and models with DVC, and how to commit their

Author: I did my best to not alter the tone of the docs (which I would describe as absolutely abhorrent) but couldn't bring myself to let this one go. All of my other changes though are grammatical corrections or at least increase precision (leaving less room for misinterpretation).

> We've learned how to _track_ data and models with DVC, and how to commit their

Reviewer: The original author here was @shcheklein if I remember correctly but of course others reviewed this at the time and passed the process. I think we need to trust our own processes, so far we have good results (esp. the GS is very popular and never had a complaint on tone, well we have one now).

But any improvement suggestions that can be explained are def. welcome. TBH it's still unclear to me what the reason is for this change. We have playful language here and there throughout the website. Hard to tell whether it has a negative impact but I doubt it. Could be wrong

> We've learned how to _track_ data and models with DVC, and how to commit their

Reviewer: BTW the change can def stay now since it's done. I'm just saying that we should try to focus on more impactful things that aren't subjective in nature, esp. for existing writing tone and style. Thanks

> We've learned how to _track_ data and models with DVC, and how to commit their

Author: marking as resolved; continuing discussion about style below in other comments.

versions to Git. The next questions are: How can we _use_ these artifacts
outside of the project? How do we download a model to deploy it? How to download
a specific version of a model? Or reuse datasets across different projects?

> These questions tend to come up when you browse the files which DVC saves to

> > remote storage (e.g.

Reviewer: On `,` -> `(` it's good that you read and review all this very closely but does it really make a difference? 🙂 I don't mind either way, I'd just try to avoid restyling existing text (no point probably: we'll always have slightly different styles per author anyway).

> > remote storage (e.g.

Author: this isn't stylistic - this is purely about readability. "Which" is more precise, and the parentheses in this case demarcate an aside, replacing a comma splice. Technically the comma could've been replaced with a semicolon instead.

> > remote storage (e.g.

Reviewer: "Which" doesn't sound correct here. As I mentioned, the clause is restrictive https://www.merriam-webster.com/words-at-play/when-to-use-that-and-which

Also, `, e.g.` is very common, at least in US academia in my experience. See https://english.stackexchange.com/questions/415616/proper-placement-of-e-g nothing wrong with it.

Please let's not focus on style. If there are big obvious non subjective errors in writing then absolutely let's fix them when we see them, otherwise we're going to spend most of our time arguing grammatical forms and we're not a publisher.

> > remote storage (e.g.

Author: > "Which" doesn't sound correct here.

Ah MW frequently riles by UK-centric sensibilities (and also note that the MW link itself acknowledges that "the best writers seem to disregard" the rule). I follow a different rule.

However I think consistency is paramount as since the entirety of the docs is very Americaneseishy I'm happy to stick to MW.

But... I think MW is consistent with my edit in this case. The MW article's conclusion is unrestricted -> which; restricted -> personal style choice.

This particular sentence is bordering on unrestricted so really "which" is technically safest in all cases :)

> Also, `, e.g.` is very common

Nothing wrong with `, e.g.` - rather I was placing the entire following sentiment into parens.

Anyway I think it's worth on occasion debating nuances lest future docs suffer. It's not about this PR, but about the millions to come.

> > remote storage (e.g.

Reviewer: > "the best writers seem to disregard" the rule.) I follow a different rule.

Sure. Just again, no strong reason to change, it's subjective, etc. I don't have a strong opinion on the specific text. I just think the best route here is to let docs have different styles per author, it's prob OK and mainly we don't have the capacity to have this kind of discussion super often but def appreciate it now and then as you mentioned 🙂 

> consistency is paramount

That too! But on the other hand if you go for new docs next don't worry about sounding Americaneseishy 🇬🇧 

> restricted -> personal style choice

That would mean leave the existing style IMO. But again whatevs, let's move on.

> `s3://dvc-public/remote/get-started/fb/89904ef053f04d64eafcc3d70db673` 😱
> instead of the original filename such as `model.pkl` or `data.xml`).

Read on or watch our video to see how to find and access models and datasets
with DVC.

https://youtu.be/EE7Gk84OZY8

Remember those `.dvc` files `dvc add` generates? Those files (and `dvc.lock`,

> which we'll cover later), have their history in Git. DVC's remote storage config

Reviewer: In this case ", which" is better indeed. Let's also remove this comma though?
```suggestion
which we'll cover later) have their history in Git. DVC's remote storage config
```

Author: yes missed this

# Get Started: Data and Model Access

We've learned how to _track_ data and models with DVC, and how to commit their
versions to Git. The next questions are: How can we _use_ these artifacts
outside of the project? How do we download a model to deploy it? How to download
a specific version of a model? Or reuse datasets across different projects?

> > These questions tend to come up when you browse the files which DVC saves to

Reviewer: Sorry, gonna go for "keep Americaneishy restrictive clause rule" on this one for now:
```suggestion
> These questions tend to come up when you browse the files that DVC saves to
```
All other changes to this note are great!

> remote storage (e.g.
> `s3://dvc-public/remote/get-started/fb/89904ef053f04d64eafcc3d70db673` 😱

> > instead of the original filename such as `model.pkl` or `data.xml`).

Reviewer: For consistency (we have it separate in most docs):
```suggestion
> instead of the original file name such as `model.pkl` or `data.xml`).
```
> Committing this, please `git pull`.

## Import file or directory

`dvc import` also downloads any file or directory, while also creating a `.dvc`

> file which can be saved in the project:

Reviewer: 👍. I'd even add parentheses since this part is not shown in the example below:
```suggestion
file (which can be saved in the project):
```

This is similar to `dvc get` + `dvc add`, but the resulting `.dvc` files
includes metadata to track changes in the source repository. This allows you to

> bring in changes from the data source later using `dvc update`.

Reviewer: Why? "source later using" could use a pause.

Author: basically a comma splice. Could parenthesise instead...

```suggestion
bring in changes from the data source later (using `dvc update`).
```

but actually also sounds better without any pause.

Reviewer: Sorry again caught up with way too much recently.

Reviewer: So again, you may be right (I disagree) but this one change seems stylistic i.e. could prob work one way or the other. Let's try to avoid such changes?

The changes where you rephrase things for clarity are much more impactful I think.

> doesn't actually contain a `get-started/data.xml` file. Like `dvc get`,
> `dvc import` downloads from [remote storage](/doc/command-reference/remote).

`.dvc` files created by `dvc import` have special fields — such as the data

> source `repo`, and `path` (under `deps`):

Reviewer: Hmm, I don't think dash works here.

Author: It's either a dash or nothing. A colon would be excessive and the comma splice is wrong. Brackets would look odd since there'd be nested parentheses.

Reviewer: Long dashes are usually for jumping ahead in the story, which isn't the case here. Instead, here we want the reader to notice certain fields in the snippet below. https://www.dummies.com/education/language-arts/grammar/how-to-use-long-and-short-dashes-in-your-writing/ (BTW we may not be using this right everywhere ATM.)

Comma is very common before "such as" https://www.grammarly.com/blog/such-as-comma/ (again for nonrestrictive clauses, such as this one)

Author: argh same MW-inspired point :)

The problem with this sentence is that `(),:` are all already present.

Reviewer: On this one I do have a not-so-weak opinion I must say. I see `blah blah — bleh:` ending in `:` and think something is wrong. No problem in having 2 commas in a sentence, I think?

Author: btw just looked at your link (this time dummies.com rather than MW) and once again think it agrees with me. Actually I agree completely with it: "the curator painted the dinosaur orange — everyone hates the color" is wrong and "the curator painted the dinosaur orange — a color hated by everyone" is correct. This is the same case.

I would not use a comma as a pause just before a comma-separated list as that would confuse things. "... have special fields, such as X, and Y" could be interpreted as "have 1. special fields (such as X), and 2. Y."

A dramatic pause (--) is absolutely essential here to show that "such as" applies to both X and Y. Otherwise in most cases I agree comma before "such as" is fine.

Reviewer: > I would not use a comma as a pause

OK you should use your own style of course. But this text is already published. What problem are we fixing?

> fields, such as X, and Y" could be interpreted ...

I see. In this case I think the improvement would be to remove the comma between "`repo`, and `path`"

Author: done

> `dvc import` downloads from [remote storage](/doc/command-reference/remote).

`.dvc` files created by `dvc import` have special fields, such as the data
source `repo` and `path` (under `deps`):

> ```git
> + deps:
> + - path: get-started/data.xml
> +   repo:

Reviewer: Also, a full fix here (including indentation) would require removing those spaces like:

```suggestion
```git
+deps:
+- path: get-started/data.xml
+  repo:
```
but there are 2 more lines that I can't suggest an edit for from GH.

Reviewer: Extracted all these (and more) to #2462 to simplify this PR a little, please review/approve if possible @casperdcl .

Reviewer: OK that PR is in, want to merge `master`? 🙂

import dvc.api

with dvc.api.open(
    'get-started/data.xml',
    repo='https://github.com/iterative/dataset-registry'
) as fd:

>     # fd is a file descriptor which can be used here

Reviewer: Why remove the "processed normally" part? And period
```suggestion
    # fd is a file descriptor, which can be processed normally.
```

Author: honestly this just need to say `fd is a file descriptor`. A code comment isn't meant to be prosaic.

Reviewer: It's not a regular code commit you would see in the actual code. The actual code would just use `fd`. This is meant to replace multiple lines of code that are irrelevant for explaining the DVC API. The point is now you have a file descriptor that you can use normally, as you would use any file descriptor.

That said I'd be OK with just removing the 2nd half since it's kind of obvious for anyone who knows what a FD is what you'd do with it. But I fear again we're not focusing on impactful things.

Author: changed
